File: src/test/java/edu/hm/hafner/analysis/IssueAssert.java
package edu.hm.hafner.analysis;

import java.util.Objects;

import org.assertj.core.api.AbstractAssert;

/**
 * Custom assertions for {@link Issue}.
 *
 * @author Joscha Behrmann
 */
@SuppressWarnings("NonBooleanMethodNameMayNotStartWithQuestion")
public class IssueAssert extends AbstractAssert<IssueAssert, Issue> {
    private static final String EXPECTED_BUT_WAS_MESSAGE = "%nExpecting:%n <%s>%nto be equal to:%n <%s>%nbut was not.";

    public IssueAssert(final Issue actual) {
        super(actual, IssueAssert.class);
    }

    public static IssueAssert assertThat(final Issue actual) {
        return new IssueAssert(actual);
    }

    public IssueAssert hasFileName(final String fileName) {
        isNotNull();

        if (!Objects.equals(actual.getFileName(), fileName)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "fileName", fileName);
        }

        return this;
    }

    public IssueAssert hasLineStart(final int lineStart) {
        isNotNull();

        if (actual.getLineStart() != lineStart) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "lineStart", lineStart); }

        return this;
    }

    public IssueAssert hasLineEnd(final int lineEnd) {
        isNotNull();

        if (actual.getLineEnd() != lineEnd) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "lineEnd", lineEnd);
        }

        return this;
    }

    public IssueAssert hasColumnStart(final int columnStart) {
        isNotNull();

        if (actual.getColumnStart() != columnStart) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "columnStart", columnStart);
        }

        return this;
    }

    public IssueAssert hasColumnEnd(final int columnEnd) {
        isNotNull();

        if (actual.getColumnEnd() != columnEnd) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "columnEnd", columnEnd);
        }

        return this;
    }

    public IssueAssert hasCategory(final String category) {
        isNotNull();

        if (!actual.getCategory().equals(category)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "category", category);
        }

        return this;
    }

    public IssueAssert hasType(final String type) {
        isNotNull();

        if (!actual.getType().equals(type)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "type", type);
        }

        return this;
    }

    public IssueAssert hasPackageName(final String packageName) {
        isNotNull();

        if (!actual.getPackageName().equals(packageName)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "packageName", packageName);
        }

        return this;
    }

    public IssueAssert hasPriority(final Priority priority) {
        isNotNull();

        if (actual.getPriority() != priority) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "priority", priority);
        }

        return this;
    }

    public IssueAssert hasMessage(final String message) {
        isNotNull();

        if (!actual.getMessage().equals(message)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "message", message);
        }

        return this;
    }

    public IssueAssert hasDescription(final String description) {
        isNotNull();

        if (!actual.getDescription().equals(description)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "description", description);
        }

        return this;
    }

    public IssueAssert hasFingerprint(final String fingerprint) {
        isNotNull();

        if (!actual.getFingerprint().equals(fingerprint)) {
            failWithMessage(EXPECTED_BUT_WAS_MESSAGE, "fingerprint", fingerprint);
        }

        return this;
    }
}

File: src/test/java/edu/hm/hafner/analysis/IssueAssert.kt
package edu.hm.hafner.analysis

import java.util.Objects
import java.util.UUID

import org.assertj.core.api.AbstractAssert

open class IssueAssert(actual: Issue) : AbstractAssert<IssueAssert, Issue>(actual, IssueAssert::class.java) {

    fun hasFileName(fileName: String): IssueAssert {
        isNotNull

        if (actual.fileName != fileName) {
            failWithMessage("Expected issue's fileName to be <%s> but was <%s>", fileName, actual.fileName)
        }

        return this
    }

    fun hasCategory(category: String): IssueAssert {
        isNotNull

        if (actual.category != category) {
            failWithMessage("Expected issue's category to be <%s> but was <%s>", category, actual.category)
        }

        return this
    }

    fun hasType(type: String): IssueAssert {
        isNotNull

        if (actual.type != type) {
            failWithMessage("Expected issue's type to be <%s> but was <%s>", type, actual.type)
        }

        return this
    }

    fun hasPriority(priority: Priority): IssueAssert {
        isNotNull

        if (actual.priority != priority) {
            failWithMessage("Expected issue's priority to be <%s> but was <%s>", priority, actual.priority)
        }

        return this
    }

    fun hasMessage(message: String): IssueAssert {
        isNotNull

        if (actual.message != message) {
            failWithMessage("Expected issue's message to be <%s> but was <%s>", message, actual.message)
        }

        return this
    }

    fun hasDescription(description: String): IssueAssert {
        isNotNull

        if (actual.description != description) {
            failWithMessage("Expected issue's description to be <%s> but was <%s>", description, actual.description)
        }

        return this
    }

    fun hasPackageName(packageName: String): IssueAssert {
        isNotNull

        if (actual.packageName != packageName) {
            failWithMessage("Expected issue's packageName to be <%s> but was <%s>", packageName, actual.packageName)
        }

        return this
    }

    fun hasLineStart(lineStart: Int): IssueAssert {
        isNotNull

        if (actual.lineStart != lineStart) {
            failWithMessage("Expected issue's lineStart to be <%s> but was <%s>", lineStart, actual.lineStart)
        }

        return this
    }

    fun hasLineEnd(lineEnd: Int): IssueAssert {
        isNotNull

        if (actual.lineEnd != lineEnd) {
            failWithMessage("Expected issue's lineEnd to be <%s> but was <%s>", lineEnd, actual.lineEnd)
        }

        return this
    }

    fun hasColumnStart(columnStart: Int): IssueAssert {
        isNotNull

        if (actual.columnStart != columnStart) {
            failWithMessage("Expected issue's columnStart to be <%s> but was <%s>", columnStart, actual.columnStart)
        }

        return this
    }

    fun hasColumnEnd(columnEnd: Int): IssueAssert {
        isNotNull

        if (actual.columnEnd != columnEnd) {
            failWithMessage("Expected issue's columnEnd to be <%s> but was <%s>", columnEnd, actual.columnEnd)
        }

        return this
    }

    fun hasId(uuid: UUID): IssueAssert {
        isNotNull

        if (actual.id != uuid) {
            failWithMessage("Expected issue's id to be <%s> but was <%s>", uuid, actual.id)
        }

        return this
    }

    fun hasFingerPrint(fingerPrint: String): IssueAssert {
        isNotNull

        if (actual.fingerprint != fingerPrint) {
            failWithMessage("Expected issue's fingerPrint to be <%s> but was <%s>", fingerPrint, actual.fingerprint)
        }

        return this
    }

    companion object {

        fun assertThat(actual: Issue): IssueAssert {
            return IssueAssert(actual)
        }
    }
}